liquid-compiler/src/filter.rs

```diff
@@ -1,101 +1,254 @@
+    pub is_optional: bool,
 }
 
-impl<T> FilterValueClone for T
-where
-    T: 'static + FilterValue + Clone,
-{
-    fn clone_box(&self) -> Box<FilterValue> {
-        Box::new(self.clone())
-    }
-}
-
-impl Clone for Box<FilterValue> {
-    fn clone(&self) -> Box<FilterValue> {
-        self.clone_box()
-    }
+/// A trait that holds the information of the parameters of a filter.
+///
+/// All structs that implement `FilterParameters` must implement this.
+/// This is actually automatically implemented with `#[derive(FilterParameters)]`.
+///
+/// This trait allows `FilterReflection` macro to extract the parameters information
+/// from a given `FilterParameters` structure.
+pub trait FilterParametersReflection {
+    fn positional_parameters() -> &'static [ParameterReflection];
+    fn keyword_parameters() -> &'static [ParameterReflection];
 }
 
-/// Function signature that can act as a `FilterValue`.
-pub type FnFilterValue = fn(&Value, &[Value]) -> FilterResult;
-
-#[derive(Clone)]
-struct FnValueFilter {
-    filter: FnFilterValue,
+/// A trait that holds the information of a filter about itself, such as
+/// its name, description and parameters.
+///
+/// All structs that implement `ParseFilter` must implement this.
+///
+/// # Deriving
+///
+/// This trait may be derived with `liquid-derive`'s `#[derive(FilterReflection)]`. However,
+/// it is necessary to use the `#[filter(...)]`  helper attribute. See documentation on
+/// `liquid-derive` for more information.
+pub trait FilterReflection {
+    fn name(&self) -> &'static str;
+    fn description(&self) -> &'static str;
+
+    fn positional_parameters(&self) -> &'static [ParameterReflection];
+    fn keyword_parameters(&self) -> &'static [ParameterReflection];
 }
 
-impl FnValueFilter {
-    fn new(filter: FnFilterValue) -> Self {
-        Self { filter }
-    }
+/// A trait that declares and holds the parameters of a filter.
+///
+/// Provides `from_args`, to construct itself from `FilterArguments` (parses the arguments)
+/// and `evaluate`, to construct its evaluated counterpart (evaluates the arguments).
+///
+/// # Deriving
+///
+/// The whole point of this structure is to facilitate the process of deriving a filter.
+/// Thus, this trait and all traits it requires may be derived with `#[derive(Debug, FilterParameters)]`.
+///
+/// See documentation for `FilterParameters` macro on `liquid-derive` for more information.
+pub trait FilterParameters<'a>: Sized + FilterParametersReflection + Debug + Display {
+    type EvaluatedFilterParameters;
+    fn from_args(args: FilterArguments) -> Result<Self>;
+    fn evaluate(&'a self, context: &'a Context) -> Result<Self::EvaluatedFilterParameters>;
 }
 
-impl FilterValue for FnValueFilter {
-    fn filter(&self, input: &Value, arguments: &[Value]) -> FilterResult {
-        (self.filter)(input, arguments)
-    }
+/// Structure that holds the unparsed arguments of a filter, both positional and keyword.
+pub struct FilterArguments<'a> {
```

Comment:
Thankfully not seeing us evolve this struct, so probably fine as-is?

Reply:
I'd say so, especially given that in the #301 draft this struct wouldn't even exist. The two type of arguments would be directly in the function declaration:
```
trait FilterParser: FilterReflection {
    fn parse(&self, Vec<Expression>, HashMap<&str, Expression>) -> Box<dyn Filter>;
}
```
The value I see in this struct is actually more related to organize the function declaration and link positional and keyword arguments, in the sense that they are both just arguments and are mostly (always?) passed in to the functions together. So I'm not really seeing it being extended.